tc/core/check.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <cstdint>
+#include <iosfwd>
+#include <sstream>
+#include <string>
+
+#include "tc/core/utils/type_traits.h"
+
+#define TC_CHECK(condition) \
+  tc::detail::tc_check(static_cast<bool>(condition), __FILE__, __LINE__)
+#define TC_CHECK_EQ(x, y) tc::detail::tc_check_eq(x, y, __FILE__, __LINE__)
+#define TC_CHECK_NE(x, y) tc::detail::tc_check_ne(x, y, __FILE__, __LINE__)
+#define TC_CHECK_LT(x, y) tc::detail::tc_check_lt(x, y, __FILE__, __LINE__)
+#define TC_CHECK_GT(x, y) tc::detail::tc_check_gt(x, y, __FILE__, __LINE__)
+#define TC_CHECK_GE(x, y) tc::detail::tc_check_ge(x, y, __FILE__, __LINE__)
+#define TC_CHECK_LE(x, y) tc::detail::tc_check_le(x, y, __FILE__, __LINE__)
+
+namespace tc {
+
+namespace detail {
+class Checker {
+ public:
+  Checker(bool condition, std::string location, std::string baseErrorMsg);
+  ~Checker() noexcept(false);
+
+  template <typename T>
+  typename std::enable_if<!tc::is_std_container<T>::value, Checker&>::type
+  operator<<(const T& msg) {
+    std::stringstream ss;
+    ss << additionalMsg_ << msg;
+    additionalMsg_ = ss.str();
+    return *this;
+  }
+
+  template <typename C>
+  typename std::enable_if<tc::is_std_container<C>::value, Checker&>::type
+  operator<<(const C& msg) {
+    std::stringstream ss;
+    ss << additionalMsg_;
+    for (const auto& x : msg) {
+      ss << x << ',';
+    }
+    additionalMsg_ = ss.str();
+    if (msg.begin() != msg.end()) {
+      additionalMsg_.pop_back();
+    }
+    return *this;
+  }
+
+ private:
+  bool condition_;
+  std::string location_;
+  std::string baseErrorMsg_;
+  std::string additionalMsg_;
```

Comment:
Why do you need to make a distinction between these two types of messages?

Reply:
There's no need. I'll merge them.

Reply:
On a second thought there is one use case:
```
TC_CHECK_X(...) << "A" << "B";
```

What I initially implemented will print the following:
```
Check failed [file:lineno] "what failed": AB
```
The `"what failed"` depends on the type of check, e.g. for `TC_CHECK_EQ(1,2)` it would be `1 not equal to 2`. This part of the output is always present.

Extra information (passed with `<<`) is optional. The reason I keep track of it in a separate string is to figure out if and where the separator (`:`) and the space before it  should be printed. 

There are two options:
1. The optional/extra information is maintained on a separate string.
2. The space and separator is appended on the first call to `operator<<` and there is extra state that keeps track of it (e.g., a boolean).